ext/drreg/drreg.c

```diff
@@ -1173,11 +1173,16 @@ drreg_statelessly_restore_app_value(void *drcontext, instrlist_t *ilist, reg_id_
          */
 #ifdef X86
     if (reg != DR_REG_NULL && pt->aflags.xchg == reg) {
+        /* drreg_restore_app_value invokes drreg_move_aflags_from_reg with stateful=false
+         * above. This spills aflags to some slot, regardless of whether they are live or
+         * not. Also, even though aflags are spilled to a slot, pt->aflags.xchg is not
+         * cleared. Below, we restore aflags to xax, and forget the spill slot.
+         */
         ASSERT(pt->aflags.slot != MAX_SPILLS, "Aflags slot not reserved");
         pt->slot_use[pt->aflags.slot] = DR_REG_XAX; /* appease assert */
         restore_reg(drcontext, pt, DR_REG_XAX, pt->aflags.slot, ilist, where_respill,
-                    false);
-        pt->slot_use[pt->aflags.slot] = DR_REG_NULL;
+                    /*release=*/true);
+        reset_aflags_spill_slot(pt);
```

Comment:
This doesn't make sense to me: this is a stateless change; there is an assert on line 1181 that aflags.slot is set, so clearing it here will cause that to fire if this routine is called again.

Also, passing release=true is duplicating the clear of slot_use in reset_aflags_spill_slot.  And again, it feels wrong to have release=true for a stateless change.  I'm not sure I remember all the invariants: but shouldn't slot_use[aflags.slot] always be set if aflags are reserved and we should not give up that slot, or else someone else could take it and we'd have aflags.slot pointing to someone else's slot?  If for aflags-in-xax we don't normally have a slot: then why is the assert at line 1181 there?

Reply:
> there is an assert on line 1181 that aflags.slot is set, so clearing it here will cause that to fire if this routine is called again.

Note that when we enter `drreg_statelessly_restore_app_value`, `aflags.slot` is **not** already set. It is set by `drreg_restore_app_value` on L1166, and then cleared at L1185. So, in the second call to `drreg_statelessly_restore_app_value`, it will be spilled again before we get to this assert.

I modified the test to call `drreg_statelessly_restore_app_value` twice consecutively, and it worked fine. In the bb, I see two pairs of spills/restores, as expected:

```
 +28   L3 @0x00007fa0fc407cf8  48 c7 c3 01 00 00 00 mov    $0x0000000000000001 -> %rbx
 +35   m4 @0x00007fa0fc407fe0  65 48 a3 e8 00 00 00 mov    %rax -> %gs:0x000000e8[8byte]
                               00 00 00 00
 +46   m4 @0x00007fa0fc408060  65 48 a1 e8 00 00 00 mov    %gs:0x000000e8[8byte] -> %rax
                               00 00 00 00
 +57   m4 @0x00007fa0fc3a71a8  65 48 a3 e8 00 00 00 mov    %rax -> %gs:0x000000e8[8byte]
                               00 00 00 00
 +68   m4 @0x00007fa0fc3a84d0  65 48 a1 e8 00 00 00 mov    %gs:0x000000e8[8byte] -> %rax
                               00 00 00 00
 +79   L3 @0x00007fa0fc407d78  48 c7 c3 02 00 00 00 mov    $0x0000000000000002 -> %rbx
```

Will update the test in the next push.

> And again, it feels wrong to have release=true for a stateless change.

We are not modifying any relevant drreg state (state that existed before the call to `drreg_statelessly_restore_app_value`). We are not releasing any slot that was in use prior to the call. The spill of aflags to a slot and its subsequent restore are local to `drreg_statelessly_restore_app_value` and is required so that we can restore xax between `where_restore` and `where_respill`.

> Also, passing release=true is duplicating the clear of slot_use in reset_aflags_spill_slot. 

Yes, passing `release=true` is redundant. But that's true for today. Tomorrow, if someone adds additional bookkeeping to update in restore_reg, we want this to still work.  As we really do intend to release the slot here, I figured its better to do the redundant thing.

> shouldn't slot_use[aflags.slot] always be set if aflags are reserved and we should not give up that slot, or else someone else could take it and we'd have aflags.slot pointing to someone else's slot

If aflags are in xax, then we don't want to hold any slot (as an optimisation). In this case, `aflags.slot = MAX_SPILLS` so we wouldn't try to read that slot.

Q: Based on the documentation of `drreg_statelessly_restore_app_value`,  I assumed that there are no app instrs between `where_respill` and `where_restore`. Is that right?
```
This routine is meant for use with instrumentation that uses separate control flow
 * paths, such as a fastpath and a slowpath, where the slowpath needs access to the
 * full application state yet must retain scratch register parity with the fastpath.
 * 
```

If not then the aflags slot used temporarily here may be overwritten by instrumentation added on the intervening app instr. Which means it's better to lock in a slot for aflags whenever it is reserved.

> If for aflags-in-xax we don't normally have a slot: then why is the assert at line 1181 there?

This is normally true, yes. But (as described above) in case of `drreg_statelessly_restore_app_value`, we have `aflags.xchg=DR_REG_XAX` (even though xax holds its native value at this point) and `aflags.slot!=MAX_SPILLS` both. The slot was reserved by `drreg_statelessly_restore_app_value` at L1166. This seemingly inconsistent intermediate state is local to `drreg_statelessly_restore_app_value`.

Reply:
> Note that when we enter drreg_statelessly_restore_app_value, aflags.slot is not already set. It is set by drreg_restore_app_value on L1166

Ah, that's what I was missing.  So if aflags.xchg is set and aflags are in xax, we don't have a slot reserved.  Let me re-examine the code w/ that info.

Reply:
> Q: Based on the documentation of drreg_statelessly_restore_app_value, I assumed that there are no app instrs between where_respill and where_restore. Is that right?

This was added primarily for Dr. Memory for the still-not-finished attempt to replace its existing scratch register management system with drreg.  Here are the two calls to this API function:
https://github.com/DynamoRIO/drmemory/blob/i1795-use-drreg/drmemory/slowpath.c#L1084

The post_jmp there is across a "lean call".  So no app instrs in between.  I think most use cases would be similar: internal control flow added for different instrumentation cases, all handling a single app instr,

Reply:
I'm still a bit worried about the tmp aflags slot colliding with some drreg reservation by instrumentation between `where_restore` and `where_respill`, which seems possible right? Or should we just document that there shouldn't be any?

Reply:
I mean, should we just document that a client shouldn't add any drreg reservation between `where_restore` and `where_respill`?

Reply:
It is up to the client to keep the reg reserved w/ a single reservation across the two points.  Could make that clearer in the docs.

Reply:
Added comment in drreg.h also to describe this limitation.

I agree on just documenting this limitation. The other option would be to let go of this optimisation of aflags not reserving a spill slot when not needed, which could be expensive.